### ISSUE TITLE
-[__NSDictionaryM setObject:forKey:]: key cannot be nil'

### ISSUE THREAD

**Issue by annamalairaj** [ra1028/DifferenceKit#13]
I am integrating differencekit with asyncdisplaykit's collectionNode in my application. I followed the instructions as mentioned in the doc and examples. But getting weird error I couldn't figure out where exactly it is happening. 
Normal reloadData() works perfectly fine. But using 
```
//self is a ASCollectionNode
self.view.reload(using: changeset) { (data) in
                    self.products = newValue
                }
```
Making the application to crash.

```
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSDictionaryM setObject:forKey:]: key cannot be nil'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000102e9b1e6 __exceptionPreprocess + 294
	1   libobjc.A.dylib                     0x0000000101feb031 objc_exception_throw + 48
	2   CoreFoundation                      0x0000000102edb0bc _CFThrowFormattedException + 194
	3   CoreFoundation                      0x0000000102dae72a -[__NSDictionaryM setObject:forKey:] + 1002
	4   UIKit                               0x00000001044faf07 -[UICollectionView _setVisibleView:forLayoutAttributes:] + 171
	5   UIKit                               0x0000000104513dcb __71-[UICollectionView _updateWithItems:tentativelyForReordering:animator:]_block_invoke.1997 + 744
	6   UIKit                               0x0000000103aef537 +[UIView(UIViewAnimationWithBlocks) _setupAnimationWithDuration:delay:view:options:factory:animations:start:animationStateGenerator:completion:] + 560
	7   UIKit                               0x0000000103aefa0f +[UIView(UIViewAnimationWithBlocks) animateWithDuration:delay:options:animations:completion:] + 99
	8   UIKit                               0x00000001045130f5 -[UICollectionView _updateWithItems:tentativelyForReordering:animator:] + 6349
	9   UIKit                               0x000000010450ccc3 -[UICollectionView _endItemAnimationsWithInvalidationContext:tentativelyForReordering:animator:] + 17420
	10  UIKit                               0x000000010451500e -[UICollectionView _endUpdatesWithInvalidationContext:tentativelyForReordering:animator:] + 71
	11  UIKit                               0x0000000104515357 -[UICollectionView _performBatchUpdates:completion:invalidationContext:tentativelyForReordering:animator:] + 439
	12  UIKit                               0x000000010451517d -[UICollectionView _performBatchUpdates:completion:invalidationContext:tentativelyForReordering:] + 91
	13  UIKit                               0x00000001045150ff -[UICollectionView _performBatchUpdates:completion:invalidationContext:] + 74
	14  UIKit                               0x0000000104515054 -[UICollectionView performBatchUpdates:completion:] + 53
	15  AsyncDisplayKit                     0x0000000100cd3e3b -[ASCollectionView _superPerformBatchUpdates:completion:] + 571
	16  AsyncDisplayKit                     0x0000000100ce9552 __64-[ASCollectionView rangeController:updateWithChangeSet:updates:]_block_invoke + 882
	17  AsyncDisplayKit                     0x0000000100ce91ab _ZL30ASPerformBlockWithoutAnimationbU13block_pointerFvvE + 123
	18  AsyncDisplayKit                     0x0000000100ce8e83 -[ASCollectionView rangeController:updateWithChangeSet:updates:] + 3859
	19  AsyncDisplayKit                     0x0000000100df694d -[ASRangeController dataController:updateWithChangeSet:updates:] + 717
	20  AsyncDisplayKit                     0x0000000100d02bf6 __40-[ASDataController updateWithChangeSet:]_block_invoke_2.224 + 230
	21  AsyncDisplayKit                     0x0000000100dcb0b4 __30-[ASMainSerialQueue runBlocks]_block_invoke + 308
	22  libdispatch.dylib                   0x00000001075c27ab _dispatch_call_block_and_release + 12
	23  libdispatch.dylib                   0x00000001075c37ec _dispatch_client_callout + 8
	24  libdispatch.dylib                   0x00000001075ce8cf _dispatch_main_queue_callback_4CF + 628
	25  CoreFoundation                      0x0000000102e5dc99 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
	26  CoreFoundation                      0x0000000102e21ea6 __CFRunLoopRun + 2342
	27  CoreFoundation                      0x0000000102e2130b CFRunLoopRunSpecific + 635
	28  GraphicsServices                    0x000000010ae53a73 GSEventRunModal + 62
	29  UIKit                               0x0000000103a2d057 UIApplicationMain + 159
	30  Sodimac                             0x0000000100705027 main + 55
	31  libdyld.dylib                       0x0000000107640955 start + 1
	32  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```
I don't know what I am missing. Any help would be much appreciated.  Thanks in advance.

**Comment by hashemp206** [comment]
Please upload a sample app. I have used difference kit with `ASCollectionNode` with no problem

**Comment by Sameesunkaria** [comment]
@annamalairaj use `data` in place of `newValue` inside of the setData closure. A `StagedChangeset` goes through multiple stages of setting data to produce the final result. Each stage has its own sets of changes and its own `data` collection. So, just setting the final value inside the closure is bound to produce a crash.

if you cannot assign `data` to `self.products` (presumably your data model) due to mis-matched types, you can either try to map the data collection to match the type/structure for your data model (This cannot be computationally intensive, as it happens on the main thread), or consider restructuring your data model to work well with DifferenceKit.

**Comment by annamalairaj** [comment]
@hashemp206  and @Sameesunkaria Thanks for the prompt reply.  @Sameesunkaria My mistake. I set it to data only and while sharing code forget to change since I was testing by switching between normal reloadData and  differenceKit reload.

Although no luck.  :(

Here is the [project](https://drive.google.com/open?id=1FmK4irtMqgO1Yr7r36UM92zAZQKgnGtv). 
Class name 
ProductListNode - differencekit

I am not sure what have I done wrong. And the clue is if any other cell node is being returned from ProductListNode no issues observed.

**Comment by annamalairaj** [comment]
@ra1028 Thank you. I tried as you mentioned. But no luck. Still the same.
And ASCollectionNode.view conforms to ASCollectionView  so I assume I will observe the similar behaviour.

Today I have found something, it is happening for the cells that is not in visible state. 
**Observations while updating data. - pagination enabled**
1. First set of data properly updating. no crash
2. Appending second set of data. no crash
3. Appending third set of data. App crashes. As it is going out of the visible area.

**Note:** If I keep scrolling while updating the data no crash works fine.

In the project linked above, comment ProductListCellNode and return LoadMoreCellNode from ProductListNode and check my observations in case you would like to have visible proof.

@ra1028  Thank you again.

**Comment by ra1028** [comment]
@annamalairaj 
Please make sure again whether the implementation of dataSource is really correct.
If the number of elements in the data collection is incorrect, thats error will might be occur.

**Comment by annamalairaj** [comment]
Pardon me, what do you mean by number of elements in the data collection?

**Comment by ra1028** [comment]
@annamalairaj 
I checked your project code again.
Since you are using a custom collection view layout, I think the bug is probably caused by its layout or AsyncDisplayKit.
Try to call `collectionViewLayout.invalidateLayout()` before performBatchUpdates.
Then make sure the layout implementation are correctly.

**Comment by annamalairaj** [comment]
@ra1028  Your guess is correct. The problem is in the customLayout. I created this custom layout all by myself. That leads to this state.
I implemented 
```
override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? 
```
But not this method

```
override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes?
```
So what happens the layout has been returned only for the specified rect but the layout attribute is empty for indexpath which is in the invisible state. 

Thats the reason, It keeps crashing when adding new element that exceeds the bound.

You are saviour. Becuase **If the number of elements in the data collection is incorrect** this changed my focus in the other direction which leads to the solution.

Thank you a lot once again. Your helping tendency, responding swiftly on issues are much appreciated.
